### DISCUSSION TITLE
Cannot set properties of undefined (setting 'step')

### DISCUSSION THREAD

**Discussion by prshantsharma** [hakimio/ngx-ui-tour#142] [Q&A] [answered]
Hi,

I'm getting an error in browser console when I start the `TourService`. 

<img width="960" alt="image" src="https://github.com/hakimio/ngx-ui-tour/assets/91114166/527ffb10-3386-429b-90c5-6828bf75c57d">

Below is my code:

    tourSteps: IStepOption[] = [{
      anchorId: 'welcome',
      content: 'Lorem ipsum dolor.',
      title: 'Welcome'
    }];

In `ngOnInit`, I put this below line of code:

     this.tourService.initialize(this.tourSteps);

I created a function to start the TourService as below:
    
    startTour() { 
      this.tourService.start();
    }

In my HTML, I put this below snippet:

    <div tourAnchor="welcome"></div>

Also, I pasted `<tour-step-template></tour-step-template>` in `app.component.html` as you suggested in the documentation.

Now. what I'm doing wrong/missing here? Please assist!

**Comment by hakimio** [comment]
This indicates a problem with your `<tour-step-template>`. Did you import `TourMatMenuModule` in your `AppModule` and was your app updated after adding `<tour-step-template></tour-step-template>` to your app component? Try to restart the app after adding `<tour-step-template>`.
Check [this demo project](https://github.com/hakimio/ngx-ui-tour/tree/master/projects/lazy-tour-demo/src) to see a simple working example.

**Reply by prshantsharma** [reply]
Hi @hakimio, thanks for the reply.

I didn't import `TourMatMenuModule` in AppModule because I'm working with multiple modules so I have one `SharedModule` and imported `TourMatMenuModule` in my SharedModule from `ngx-ui-tour-md-menu`. And I have imported this `SharedModule` in every module of the application.

One more thing here, I'm importing `TourMatMenuModule` using `forRoot()`. If I don't use `forRoot()`, I'm getting this below error:

<img width="960" alt="image" src="https://github.com/hakimio/ngx-ui-tour/assets/91114166/1098c031-c67a-4223-a78f-4c9decb0c36f">


And if importing `TourMatMenuModule` in `AppModule`, getting this below issue in console when I start the `TourService`:

<img width="960" alt="image" src="https://github.com/hakimio/ngx-ui-tour/assets/91114166/19fafc09-6788-4e36-9108-59a80c85c716">


Now, please share your valuable thoughts what I'm doing wrong here?


*Specifications:*
Angular version is: 13.1
Node version: 14.21
ngx-ui-tour-md-menu version: 8.1.2

**Comment by hakimio** [comment]
Yes, when using old versions of `ngx-ui-tour`, you have to import `TourMatMenuModule.forRoot()` in your `AppModule`. That's stated in the "Setup" guide.
If you have `<tour-step-template>` in your `app.component.html` and `TourMatMenuModule.forRoot()` in your `app.module.ts`, one more thing you can check is that you are importing `TourService` from `ngx-ui-tour-md-menu`, not `ngx-ui-tour-core`. In your case there shouldn't be any imports from `ngx-ui-tour-core`.
Also, make sure that you are using `v8` of both `ngx-ui-tour-md-menu` and `ngx-ui-tour-core`.
If you still can't get it to work, create a GitHub repo with the issue and I'll take a look.

**Reply by Prshant-Sharma** [reply]
@hakimio, here you can find the GitHub repo [link](https://github.com/Prshant-Sharma/ui-tour/).

Hope it will help you to guide  me better. Thanks for your concern.

**Reply by hakimio** [reply]
Your `AbcModule` has to import `TourMatMenuModule`, if you want to use `tourAnchor` directive in that module.

**Reply by hakimio** [reply]
After importing `TourMatMenuModule`, your tour works fine.

**Reply by Prshant-Sharma** [reply]
@hakimio, I imported `TourMatMenuModule` from `ngx-ui-tour-md-menu` in my `AbcModule` as shown below:

<img width="960" alt="image" src="https://github.com/hakimio/ngx-ui-tour/assets/63813857/0d4507fa-13e9-4449-8a04-7e9232e08020">


but still getting this below error:

<img width="960" alt="image" src="https://github.com/hakimio/ngx-ui-tour/assets/63813857/c2843136-6303-48eb-bad9-42ea2e576eae">

**Reply by hakimio** [reply]
It shouldn't be `forRoot()`:
```ts
import {TourMatMenuModule} from 'ngx-ui-tour-md-menu';

@NgModule({
  declarations: [
    HelloComponent,
    SidebarComponent,
    WelcomeComponent,
    RegistrationComponent,
    FinishComponent
  ],
  imports: [
    CommonModule,
    AbcRoutingModule,
    TourMatMenuModule
  ]
})
export class AbcModule { }
```

**Reply by hakimio** [reply]
Here is a patch to fix your demo:
[Tour_fix_.patch](https://github.com/hakimio/ngx-ui-tour/files/12417837/Tour_fix_.patch)

**Reply by hakimio** [reply, marked as answer]
You have to import `TourMatMenuModule.forRoot()` in your `AppModule` and `TourMatMenuModule` (without `forRoot()`) in your lazy loaded `AbcModule`.

**Reply by Prshant-Sharma** [reply]
@hakimio, amazing!!!! It really worked. 
Many thanks for your efforts! 👍
